scripts/inotify.c

```diff
@@ -0,0 +1,176 @@
+#include <errno.h>
+#include <limits.h>
+#include <poll.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/inotify.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+extern char **environ;
+
+// poll() timeout for inotify events, in milliseconds;
+// run callback after this delay since the last run, even if there're no events.
+const int kPollTimeout = 2000;
+
+char buf[1024 * (sizeof(struct inotify_event) + NAME_MAX + 1)];
+
+void run_callback_internal(char** callback) {
+  fflush(stdout); // avoid mixing with callback output.
+
+  pid_t pid = fork();
+  if (pid == -1) {
+    perror("fork");
+    return; // let it retry at the next event or timeout.
+  }
+
+  if (pid == 0) {
+    execve(callback[0], callback, environ);
+    perror("execve");
+    exit(EXIT_FAILURE);
+  }
+
+  int status;
+  do {
+    int w = waitpid(pid, &status, 0);
+    if (w == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+      perror("waitpid");
+      exit(EXIT_FAILURE); // shouldn't happen; exit & don't risk leaking zombies.
+    }
+  } while (!WIFEXITED(status) && !WIFSIGNALED(status));
+
+  printf("inotify: %s exited with status %d (%s: %d)\n", callback[0], status,
+         WIFEXITED(status) ? "exit" : WIFSIGNALED(status) ? "signal" : "unknown",
+         WIFEXITED(status) ? WEXITSTATUS(status) : WIFSIGNALED(status) ? WTERMSIG(status) : status);
+
+  if (WIFEXITED(status) && WEXITSTATUS(status) == EXIT_SUCCESS) {
+    exit(EXIT_SUCCESS);
+  }
+}
+
+// run_callback runs the callback if always is true, or it has been kPollTimeout
+// milliseconds since 'last', which is updated after callback runs.
+void run_callback(char** callback, struct timespec *last, bool always) {
+  if (!always) {
```

Comment:
I'm mildly concerned that we need to maintain another chunk of code that hands such corner case. I know this may not be easy - could we add unit tests?